pkg/verifier/safe_delete.go

```diff
@@ -31,23 +35,33 @@ func SafeDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bac
 		return errors.Errorf("%s dir seems to exists in backup bucket. Remove this block manually if you are sure it is safe to do", id)
 	}
 
-	tempdir, err := ioutil.TempDir("", "safe-delete")
-	if err != nil {
-		return err
+	if tempdir == "" {
+		// tempdir unspecified, create one
+		tempdir, err = ioutil.TempDir("", fmt.Sprintf("safe-delete-%s", id.String()))
+		if err != nil {
+			return err
+		}
 	}
 	dir := filepath.Join(tempdir, id.String())
-	err = os.Mkdir(dir, 0755)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := os.RemoveAll(tempdir); err != nil {
-			level.Warn(logger).Log("msg", "failed to delete dir", "dir", tempdir, "err", err)
+
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
```

Comment:
we're swallowing any error that is not "IsNotExist" here

Reply:
I think we should just do here `os.Mkdir` and ignore the error if it is saying that the directory already exists. Otherwise, we will have racy behaviour here (of course, there's barely any chance that someone will run two of these at the same time but still)

Reply:
Yeah, good suggestions!

Reply:
I added some additional logic here to handle any other possible error from `os.Stat()`.  

The pre-existence of the directory triggers the code path where we can avoid having to re-download a possibly large TSDB block again.  Races with the `os.Mkdir()` are not anticipated as if the temporary directory isn't supplied `ioutil.TempDir()`is used to generate a value.  When we do supply `tempdir` in `pkg/verifier/index_issue.go` is it also generated with `ioutil.Tempdir()`:

```
tmpdir, err := ioutil.TempDir("", fmt.Sprintf("index-issue-block-%s-", id))
```